core/modules/nat.cc

```diff
@@ -0,0 +1,239 @@
+                              const NAT::Flow &flow) {
+  struct udp_hdr *udp = reinterpret_cast<struct udp_hdr *>(l4);
+  struct icmp_hdr *icmp = reinterpret_cast<struct icmp_hdr *>(l4);
+
+  ip->src_addr = flow.src_ip;
+  ip->dst_addr = flow.dst_ip;
+
+  switch (flow.proto) {
+    case 0x06:
+    case 0x11:
+      udp->src_port = flow.src_port;
+      udp->dst_port = flow.dst_port;
+      break;
+    case 0x01:
+      switch (icmp->icmp_type) {
+        case 0:
+        case 8:
+        case 13:
+        case 15:
+        case 16:
+          icmp->icmp_ident = flow.icmp_ident;
+          break;
+        default:
+          VLOG(1) << "Unknown icmp_type: " << icmp->icmp_type;
+      }
+    default:
+      VLOG(1) << "Unknown next_proto_id: " << ip->next_proto_id;
+  }
+  compute_cksum(ip, l4);
+}
+
+void NAT::ProcessBatch(bess::PacketBatch *batch) {
+  gate_idx_t out_gates[bess::PacketBatch::kMaxBurst];
+  gate_idx_t incoming_gate = get_igate();
+
+  int cnt = batch->cnt();
+
+  for (int i = 0; i < cnt; i++) {
+    bess::Packet *pkt = batch->pkts()[i];
+
+    // By default, pass packet
+    out_gates[i] = incoming_gate;
+
+    struct ether_hdr *eth = pkt->head_data<struct ether_hdr *>();
+    struct ipv4_hdr *ip = reinterpret_cast<struct ipv4_hdr *>(eth + 1);
+    int ip_bytes = (ip->version_ihl & 0xf) << 2;
+    struct udp_hdr *udp = reinterpret_cast<struct udp_hdr *>(
+        reinterpret_cast<uint8_t *>(ip) + ip_bytes);
+
+    // L4 protocol must be TCP, UDP, or ICMP
+    if (ip->next_proto_id != 0x06 && ip->next_proto_id != 0x11 &&
+        ip->next_proto_id != 0x01) {
+      out_gates[i] = DROP_GATE;
+      continue;
+    }
+
+    NAT::Flow flow = parse_flow(ip, udp);
+    uint64_t now = tsc_to_ns(rdtsc());
+
+    auto hash_it = flow_hash_.find(flow);
+
+    if (hash_it != flow_hash_.end()) {
+      const NAT::Flow &translated_flow = (incoming_gate == 0)
+                                             ? hash_it->second.external_flow
+                                             : hash_it->second.internal_flow;
+
+      if (now - hash_it->second.time < TIME_OUT) {
+        // Entry exists and does not exceed timeout
+        hash_it->second.time = now;
+        stamp_flow(ip, udp, translated_flow);
+        continue;
+      } else {
+        // Reclaim expired record
+        available_ports_.push_back(hash_it->second.port);
+        flow_hash_.erase(hash_it);
+        flow_hash_.erase(translated_flow.reverse_flow());
+      }
+    }
+
+    if (incoming_gate == 1) {
+      // Flow from external network, drop.
+      out_gates[i] = DROP_GATE;
+      continue;
+    }
+
+    const auto rule_it = std::find_if(
+        rules_.begin(), rules_.end(),
+        [&ip](const NATRule &rule) { return rule.first.Match(ip->src_addr); });
+    if (rule_it == rules_.end()) {
+      // No rules found for this source IP address, drop.
+      out_gates[i] = DROP_GATE;
+      continue;
+    }
+
+    uint16_t new_port = available_ports_.back();
```

Comment:
It is randomized at the start, I agree, but if I'm reading it right, do the ports get randomized upon re-use?